decorators/classes.py

```diff
@@ -69,21 +87,140 @@ def wrapped(*args, **kwargs):
+
+    Model Map:
+    Model map is a list of tuples, where each tuple is a mapping of:
+    - Model if first item, otherwise relation
+    - URL key
+    - Field name
+    - Extra filters (Optional)
+
+    NOTE: If the first model is being provided from a previous decorator, you should
+    only provide kwargs key and nothing else.
+
+    In extra filters if current user must be used, value should be set to
+    `__current_user__`.
+
+    Example with standard usage:
+    @resolve_records_from_url_patterns(
+        (Organization, 'id', 'organization_id', {'is_active': True, 'user': '__current_user__'}),
+        ('members', 'id', 'member_id'),
+    )
+
+    Example with previous decorator:
+    @require_organization_admin()
+    @resolve_records_from_url_patterns(
+        'organization',  # organization is the kwargs key from previous decorator
+        ('members', 'id', 'member_id'),
+    """
+
+    def __init__(self, model_map, content_type='text/html'):
+        self.content_type = content_type
+        self.model_map = model_map
+
+    def __call__(self, view_fn):
+        @wraps(view_fn)
+        def wrapped(request, *args, **kwargs):
+            self.user = request.user
+
+            # If first item is just a string, it means we need to get the object
+            # from kwargs
+            if isinstance(self.model_map[0], str):
+                obj = kwargs.get(self.model_map[0])
+                model_map = self.model_map[1:]
+            else:
+                obj = None
+                model_map = self.model_map
+
+            for item in model_map:
+                model_or_relation = item[0]
+                field = item[1]
+                url_key = item[2]
+                extra_filter = len(item) > 3 and item[3] or {}
+
+                value = kwargs.get(url_key)
+
+                # Raises exception and breaks loop
+                key, obj = self._resolve_object(
+                    obj, model_or_relation, field, value, extra_filter
+                )
+
+                kwargs[key] = obj
+
+            response = view_fn(request, *args, **kwargs)
+
+            return response
+
         return wrapped
 
+    def _get_class_name(self, model=None):
+        if model is None:
+            value = self.__class__.__name__
+        elif model.__class__.__name__ == 'RelatedManager':
+            value = model.model.__qualname__
+        else:
+            value = model.__class__.__name__
+
+        return value
+
+    def _resolve_object(
+        self, obj, model_or_relation, field, value, extra_filters
+    ):
+        model = (
+            model_or_relation.objects
+            if obj is None
+            else getattr(obj, model_or_relation)
+        )
+        DoesNotExist = (
+            model.model.DoesNotExist
+            if model.__class__.__name__ == 'RelatedManager'
+            else model_or_relation.DoesNotExist
+        )
+
+        filters = extra_filters.copy()
```

Comment:
Why copy?

Reply:
Since `extra_fields` is a `dict` we are doing operations on a reference and it might actually edit the original and mess up.